test/test_binary_ufuncs.py

```diff
@@ -223,44 +223,56 @@ def test_comparison_ops_check_for_zerodim_tensor_overflow(self, device):
         with self.assertRaisesRegex(RuntimeError, 'value cannot be converted to type'):
             self.assertTrue(ts2 != t2)
 
-    # TODO: update to work on CUDA, too
-    @onlyCPU
     def test_bitwise_ops(self, device):
-        x = torch.randn(5, 5).gt(0)
-        y = torch.randn(5, 5).gt(0)
+        def check_op(x, y):
```

Comment:
I think there's a more elegant way to do this. What would you think about the following?

```
# Tests that the binary operators and, or, and xor (as well as their reflected and inplace versions) work properly (AKA &, ||, ^ and &=, |=, ^=)
@dtypes(torch.bool, torch.float32)
test_bitwise_ops(self, device, dtype):
  # Tensor x Tensor and Tensor x Scalar ops
  ops = (operator.and_, operator.iand, operator.or_, operator.ior, operator.xor, operator.ixor)
  inplace_ops = (operator.iand, operator.ior, operator.ixor)
  shapes = ((5,), (15, 15), (500, 500))

  for op, shape in product(ops, shapes):
    # Tests tensor x tensor case
    a = make_tensor(shape, device=device, dtype=dtype)
    b =  make_tensor(shape, device=device, dtype=dtype)
    a_np = a.cpu().clone().numpy()

    self.assertEqual(op(a, b), op(a_np, b_np))

    # Tests tensor x scalar case
    a = make_tensor(shape, device=device, dtype=dtype)
    b = make_tensor((), device='cpu', dtype=dtype).item()
    
    a_np = a.cpu().clone().numpy()

    self.assertEqual(op(a, b), op(a_np, b_np))
  
    # Tests scalar x tensor case (for ops which aren't inplace)
    if op in inplace_ops:
      continue

    self.assertEqual(op(b, a), op(b, a_np))
```

What would you think of an approach like this?

Reply:
cc @pmeier as an fyi to consider how we can automate this in the future

Reply:
It seems better than my implementation :smiley:. I updated the test code.

Reply:
> fyi to consider how we can automate this in the future

Not sure how we can improve on this. One thing that I was thinking on for quite a bit is a tagging system for `OpInfo`'s. Apart from the major properties (unary, binary, ...) each `OpInfo` could have a set of enums that give some additional information. If we had something like that, we could simply do `@ops([op for op in op_db if "bitwise" in op.tags])`

Reply:
Sure; typically we derive filters from the metadata. Instead of tags we could probably just add another attribute